packages/esbuild-plugin/src/index.ts

```diff
@@ -0,0 +1,214 @@
+    return vanillaExtractFilescopePlugin({ projectRoot });
+  }
+
+  return {
+    name: 'vanilla-extract',
+    setup(build) {
+      build.onResolve({ filter: /vanilla\.css\?source=.*$/ }, (args) => {
+        return {
+          path: args.path,
+          namespace: vanillaCssNamespace,
+        };
+      });
+
+      build.onLoad(
+        { filter: /.*/, namespace: vanillaCssNamespace },
+        ({ path }) => {
+          const [, source] = path.match(/\?source=(.*)$/) ?? [];
+
+          if (!source) {
+            throw new Error('No source in vanilla CSS file');
+          }
+
+          return {
+            contents: Buffer.from(source, 'base64').toString('utf-8'),
+            loader: 'css',
+          };
+        },
+      );
+
+      build.onLoad({ filter: /\.css\.(js|jsx|ts|tsx)$/ }, async ({ path }) => {
+        const result = await esbuild({
+          entryPoints: [path],
+          metafile: true,
+          bundle: true,
+          external: ['@vanilla-extract', ...externals],
+          platform: 'node',
+          write: false,
+          plugins: [vanillaExtractFilescopePlugin({ projectRoot })],
+        });
+
+        const { outputFiles } = result;
+
+        if (!outputFiles || outputFiles.length !== 1) {
+          throw new Error('Invalid child compilation');
+        }
+
+        type Css = Parameters<Adapter['appendCss']>[0];
+        const cssByFileScope = new Map<string, Array<Css>>();
+        const localClassNames = new Set<string>();
+
+        const cssAdapter: Adapter = {
+          appendCss: (css, fileScope) => {
+            if (outputCss) {
+              const fileScopeCss = cssByFileScope.get(fileScope) ?? [];
+
+              fileScopeCss.push(css);
+
+              cssByFileScope.set(fileScope, fileScopeCss);
+            }
+          },
+          registerClassName: (className) => {
+            localClassNames.add(className);
+          },
+          onEndFileScope: () => {},
+        };
+
+        setAdapter(cssAdapter);
+
+        const sourceWithBoundLoaderInstance = `require('@vanilla-extract/css/adapter').setAdapter(__adapter__);${outputFiles[0].text}`;
+
+        const evalResult = evalCode(
+          sourceWithBoundLoaderInstance,
+          path,
+          { console, __adapter__: cssAdapter },
+          true,
+        );
+
+        const cssRequests = [];
+
+        for (const [fileScope, fileScopeCss] of cssByFileScope) {
+          const css = transformCss({
+            localClassNames: Array.from(localClassNames),
+            cssObjs: fileScopeCss,
+          }).join('\n');
+          const base64Css = Buffer.from(css, 'utf-8').toString('base64');
+
+          cssRequests.push(`${fileScope}.vanilla.css?source=${base64Css}`);
+        }
+
+        const contents = serializeVanillaModule(cssRequests, evalResult);
+
+        return {
+          contents,
+          loader: 'js',
+        };
+      });
+    },
+  };
+}
+
+const stringifyExports = (value: any) =>
```

Comment:
More just marking this for future conversation, but this makes me think it might be good to have a more first-class integration API so stuff like this is handled for you.